src/udisksutabmonitor.c

```diff
@@ -0,0 +1,237 @@
+  gboolean need_reload;
+  int r;
+
+  if (cond & ~G_IO_IN)
+    return TRUE;
+
+  need_reload = FALSE;
+
+  do
+  {
+    const char *filename;
+    r = mnt_monitor_next_change (monitor->mn, &filename, NULL);
+
+    if (r == 0)
+      need_reload = TRUE;
+  } while (r == 0);
+
+  if (need_reload)
+    reload_utab_entries (monitor);
+
+  return TRUE;
+}
+
+static void
+udisks_utab_monitor_constructed (GObject *object)
+{
+  UDisksUtabMonitor *monitor = UDISKS_UTAB_MONITOR (object);
+
+  monitor->mn = mnt_new_monitor();
+  mnt_monitor_enable_userspace (monitor->mn, TRUE, NULL);
+
+  monitor->utab_channel = g_io_channel_unix_new (mnt_monitor_get_fd (monitor->mn));
+  monitor->utab_watch_source = g_io_create_watch (monitor->utab_channel, G_IO_IN);
+  g_source_set_callback (monitor->utab_watch_source, (GSourceFunc) utab_changed_event, monitor, NULL);
+  g_source_attach (monitor->utab_watch_source, g_main_context_get_thread_default ());
+  g_source_unref (monitor->utab_watch_source);
+
+  if (G_OBJECT_CLASS (udisks_utab_monitor_parent_class)->constructed != NULL)
+    (*G_OBJECT_CLASS (udisks_utab_monitor_parent_class)->constructed) (object);
+}
+
+/**
+ * udisks_utab_monitor_new:
+ *
+ * Creates a new #UDisksUtabMonitor object.
+ *
+ * Signals are emitted in the <link
+ * linkend="g-main-context-push-thread-default">thread-default main
+ * loop</link> that this function is called from.
+ *
+ * Returns: A #UDisksUtabMonitor. Free with g_object_unref().
+ */
+UDisksUtabMonitor *
+udisks_utab_monitor_new (void)
+{
+  return UDISKS_UTAB_MONITOR (g_object_new (UDISKS_TYPE_UTAB_MONITOR, NULL));
+}
+
+/**
+ * udisks_utab_monitor_get_entries:
+ * @monitor: A #UDisksUtabMonitor.
+ *
+ * Gets all /run/mounts/utab entries
+ *
+ * Returns: (transfer full) (element-type UDisksUtabEntry): A list of #UDisksUtabEntry objects that must be freed with g_list_free() after each element has been freed with g_object_unref().
+ */
+GList *
+udisks_utab_monitor_get_entries (UDisksUtabMonitor  *monitor)
+{
+  GList *ret;
+
+  g_return_val_if_fail (UDISKS_IS_UTAB_MONITOR (monitor), NULL);
+
+  udisks_utab_monitor_ensure (monitor);
+
+  ret = g_list_copy (monitor->utab_entries);
+  g_list_foreach (ret, (GFunc) g_object_ref, NULL);
+  return ret;
+}
+
+static void
+udisks_utab_monitor_ensure (UDisksUtabMonitor *monitor)
+{
+  struct libmnt_table *tb;
+  struct libmnt_iter *itr;
+  struct libmnt_fs *fs;
+  int r;
+
+  if (monitor->have_data)
+    return;
+
+  tb = mnt_new_table ();
+  r = mnt_table_parse_mtab (tb, NULL);
+
+  itr = mnt_new_iter (MNT_ITER_FORWARD);
+  while(mnt_table_find_next_fs (tb, itr, fs_has_user_options, NULL, &fs) == 0)
+  {
+    UDisksUtabEntry *entry;
+    entry = _udisks_utab_entry_new (fs);
+    monitor->utab_entries = g_list_prepend (monitor->utab_entries, entry);
+  }
+  mnt_free_iter (itr);
+
+  monitor->have_data = TRUE;
+}
+
+static void
+udisks_utab_monitor_invalidate (UDisksUtabMonitor *monitor)
```

Comment:
I believe there has to be some mutual exclusion mechanism involved here to prevent multiple threads from stepping on each others' toes when working with the same monitor instance (e.g. one using the data, the other doing the updates).